### PR TITLE
Enable autopxd to use Microsoft Visual C++ for preprocessing on Windows

### PR DESCRIPTION
**Pull request by zooba**

### CODE REVIEW

**Comment by zooba** [comment]
I'm publishing this as a draft, because I may not get time to finish it up, but wanted to share the work in case someone else can pick it up.

Basically, I wanted this to work without installing MinGW or some other build of GCC, so I added handling for when then `cpp` command is missing on Windows and made it use MSVC instead.

In case you've never seen it before, the VS discovery is how it's supposed to work. There are no suitable registry entries for VS anymore, and the vswhere tool is always in the same location. It does, however, need better handling for when the tool is missing. If the tool is missing, there are no VS installs, which would be an error here anyway! But the error message could be better - similarly if the returned JSON is an empty list (i.e. no VS install included `cl.exe`).

Also tests, of course, though it appears you don't have Windows CI set up anyway. It seems to work okay anyway though 👍

**Comment by zooba** [comment]
I just rebased and squashed everything as there's nothing interesting in my history.

I _wish_ the Windows tests could be quicker - they are embarrassingly slow. It's just the nature of launching processes on Windows. That's the big heavy process, and because it keeps shelling out to MSVC or GCC multiple times, it's inherently slow. If you'd rather not have the tests run on every PR, I understand

File: autopxd/__init__.py [hunk 1]
```diff
@@ -34,11 +36,92 @@ def ensure_binary(s, encoding="utf-8", errors="strict"):
     raise TypeError(f"not expecting type '{type(s)}'")
 
 
+def _find_cl():
+    """Use vswhere.exe to locate the Microsoft C compiler."""
+    host_platform = {
+        "X86": "X86",
+        "AMD64": "X64",
+        "ARM64": "ARM64",
+    }.get(platform.machine(), "x86")
```
Comment (elijahr) [inline autopxd/__init__.py]: Should this default to uppercase? Or should they all be lowercase?

```suggestion
    }.get(platform.machine(), "X86")
```

File: autopxd/__init__.py [hunk 2]
```diff
@@ -34,11 +36,92 @@ def ensure_binary(s, encoding="utf-8", errors="strict"):
...
+    build_platform = {
+        "X86": "x86",
+        "AMD64": "x64",
+        "ARM64": "arm64",
+    }.get(platform.machine(), "x86")
+    cmd = [
+        os.path.expandvars(r"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe"),
```
Comment (elijahr) [inline autopxd/__init__.py]: Should it be `ProgramFiles({host_platform})`?

```suggestion
        os.path.expandvars(fr"%ProgramFiles({host_platform})%\Microsoft Visual Studio\Installer\vswhere.exe"),
```

File: autopxd/__init__.py [hunk 3]
```diff
@@ -34,11 +36,92 @@ def ensure_binary(s, encoding="utf-8", errors="strict"):
...
+            *(extra_cpp_args or []),
+            source_file,
+        ]
+        with subprocess.Popen(cmd, stdout=subprocess.PIPE) as proc:
+            while proc.poll() is None:
+                result.extend(proc.communicate()[0].decode("utf-8").splitlines())
+    finally:
+        os.unlink(source_file)
+    if proc.returncode:
+        raise Exception("Invoking C preprocessor failed")
```
Comment (elijahr) [inline autopxd/__init__.py]: Maybe:

```suggestion
        raise RuntimeError("Invoking C preprocessor failed")
```

**Review by elijahr — COMMENTED** [review]
Thanks so much for contributing. I don't have Windows to test these changes with and am unfamiliar with the stack, so just have a few clarifying questions.

File: autopxd/__init__.py (hunk 2 above)
Comment (zooba) [reply autopxd/__init__.py]: Actually, the case where this needs to change is when running on a 32-bit OS (which is incredibly rare these days, but I'll add it anyway).

**Comment by elijahr** [comment]
This is great. Really appreciate it! Will publish a new release with these changes soon.

**Comment by elijahr** [comment]
This has been released in [v2.2.2](https://pypi.org/project/autopxd2/2.2.2/).

**Comment by jakirkham** [comment]
Thanks Steve! 🙏

Nice to see Windows support here 😄

**Pull request merged**